### ISSUE TITLE
Declarative vs. imperative coding style using Hooks

### ISSUE THREAD

**Issue by johnloven** [facebook/react#15494]
_If this should be asked on Stack Overflow instead, please let me know and feel free to close the issue._

Consider a component that fetches some data in a custom hook, saves the fetched data in a state hook, and notifies the user that data has been fetched using a prop callback.

Notifying the user can be done imperatively:

```js
function Component(props) {
  const [ data, setData ] = useState(null);

  useApi("/api/data")
    .then(setData)
    .then(props.onFetched);

  return dataToElements(data);
}
```

or declaratively using an effect hook:

```js
function Component(props) {
  const [ data, setData ] = useState(null);

  useApi("/api/data")
    .then(setData);

  useEffect(() => {
    if (data) {
      props.onFetched();
    }
  }, [ data ]);

  return dataToElements(data);
}
```

React seems to promote a declarative approach. But what I've found is that when components grow large and complex, using declarative effect hooks makes the flow of data and actions quite hard to follow. If you're not careful, a lot of things start to depend on a lot of other things, and the predicted results become non-intuitive and hard to wrap your head around.

I would like to know other peoples' opinions on this matter, and whether or not an imperative approach might sometimes be better.

**Comment by threepointone** [comment]
Yeah this belongs on stack overflow. That said, some quick notes - 
- you shouldn't do sideeffects in the body of a function component regardless. So your first example should be - 
```jsx
function Component(props) {
  const [ data, setData ] = useState(null);
  useEffect(() => {
    useApi("/api/data")
      .then(setData)
      .then(props.onFetched);
  })
  
  return dataToElements(data);
}
```
This particular usecase (fetching data) is common enough that Suspense (when it's released) will have a much better idiom for it.
- In your second example, I wouldn't recommend using effects to 'declaratively' align your effects like you have. The 'declarative' paradigm is specifically about component structure and passing props and events through the tree. Feel free to use 'imperative' code inside effects, that's what they're there for. 

I'm closing this issue, but if your'e interested in diving into specifics, I would would recommend reading  pretty much every post on https://overreacted.io/. This one on useEffect https://overreacted.io/a-complete-guide-to-useeffect/ is one of my favourites.

**Comment by johnloven** [comment]
Thank you @threepointone for your input, I will check out the links and probably post a similar question to Stack Overflow.

Just to clarify, the `useApi` custom hook uses an `useEffect(..., [])` hook internally so that it only fetches data on mount, so the wrapping you added is not necessary.
